File: aoc2025/python/day4.py
import time
from utilities import read_file


EMPTY = '.'
PAPER_ROLL = '@'


class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __eq__(self, other):
        return self.x == other.x and self.y == other.y

    def __hash__(self):
        return hash((self.x, self.y))

    def __repr__(self):
        return f"Point({self.x}, {self.y})"


RIGHT = Point(1, 0)
LEFT = Point(-1, 0)
DOWN = Point(0, 1)
UP = Point(0, -1)
RIGHT_DOWN = Point(1, 1)
RIGHT_UP = Point(1, -1)
LEFT_DOWN = Point(-1, 1)
LEFT_UP = Point(-1, -1)

DIRECTIONS = [LEFT, RIGHT, UP, DOWN, LEFT_DOWN, LEFT_UP, RIGHT_DOWN, RIGHT_UP]


class Grid:
    def __init__(self, filename):
        lines = read_file(filename)
        self.height = len(lines)
        self.width = len(lines[0]) if lines else 0
        self.map = [list(line) for line in lines]

    def set_object(self, point, obj):
        self.map[point.y][point.x] = obj

    def position_contains_object(self, point, obj):
        if not self.point_within_bounds(point):
            return False
        return self.map[point.y][point.x] == obj

    def point_within_bounds(self, point):
        return 0 <= point.x < self.width and 0 <= point.y < self.height


def move_point(point, direction):
    return Point(point.x + direction.x, point.y + direction.y)


def count_paper_roll_contacts(grid, point):
    count = 0
    for direction in DIRECTIONS:
        if grid.position_contains_object(move_point(point, direction), PAPER_ROLL):
            count += 1
    return count


def calculate_paper_rolls_queue(grid):
    part1_count = 0
    part2_count = 0

    stack = []
    for y in range(grid.height):
        for x in range(grid.width):
            if grid.map[y][x] == PAPER_ROLL:
                point = Point(x, y)
                connected_rolls = count_paper_roll_contacts(grid, point)
                if connected_rolls < 4:
                    part1_count += 1
                    stack.append(point)

    while stack:
        point = stack.pop()

        if grid.map[point.y][point.x] != PAPER_ROLL:
            continue

        grid.set_object(point, EMPTY)
        part2_count += 1

        for direction in DIRECTIONS:
            neighbor = move_point(point, direction)
            if grid.position_contains_object(neighbor, PAPER_ROLL):
                connected_rolls = count_paper_roll_contacts(grid, neighbor)
                if connected_rolls < 4:
                    stack.append(neighbor)

    return part1_count, part2_count


def calculate_paper_rolls_optimized(grid):
    part1_count = 0
    part2_count = 0

    mapped_paper_rolls = {}
    for y in range(grid.height):
        for x in range(grid.width):
            if grid.map[y][x] == PAPER_ROLL:
                point = Point(x, y)
                mapped_paper_rolls[point] = True
                connected_rolls = count_paper_roll_contacts(grid, point)
                if connected_rolls < 4:
                    part1_count += 1

    last_pass_count = -1
    while last_pass_count != 0:
        last_pass_count = 0

        points_to_remove = []
        for point in mapped_paper_rolls:
            connected_rolls = count_paper_roll_contacts(grid, point)
            if connected_rolls < 4:
                grid.set_object(point, EMPTY)
                points_to_remove.append(point)
                last_pass_count += 1
                part2_count += 1

        for point in points_to_remove:
            del mapped_paper_rolls[point]

    return part1_count, part2_count


def calculate_paper_rolls(grid):
    part1_count = 0
    part2_count = 0

    first_pass = True
    last_pass_count = -1

    while last_pass_count != 0:
        last_pass_count = 0

        for y in range(grid.height):
            for x in range(grid.width):
                if grid.map[y][x] != PAPER_ROLL:
                    continue

                connected_rolls = count_paper_roll_contacts(grid, Point(x, y))
                if connected_rolls < 4:
                    if first_pass:
                        part1_count += 1
                    else:
                        grid.set_object(Point(x, y), EMPTY)
                        last_pass_count += 1
                        part2_count += 1

        if first_pass:
            first_pass = False
            last_pass_count = -1

    return part1_count, part2_count


def day4():
    grid = Grid("day4.part1")

    start = time.time()
    part1_count_queue, part2_count_queue = calculate_paper_rolls_queue(grid)
    queue_time = time.time() - start

    print(f"Day 4 Queue:")
    print(f"  part1: {part1_count_queue}, part2: {part2_count_queue} ({queue_time:.4f}s)")


if __name__ == "__main__":
    day4()

File: aoc2025/javascript/day4.js
import { Grid, PAPER_ROLL, EMPTY, DIRECTIONS, Point } from './grid.js';

function movePoint(point, direction) {
  return new Point(point.x + direction.x, point.y + direction.y);
}

function countPaperRollContacts(grid, point) {
  let count = 0;

  for (const direction of DIRECTIONS) {
    if (grid.positionContainsObject(movePoint(point, direction), PAPER_ROLL)) {
      count++;
    }
  }

  return count;
}

function calculatePaperRollsQueue(grid) {
  let part1Count = 0;
  let part2Count = 0;

  const stack = [];
  for (let y = 0; y < grid.height; y++) {
    for (let x = 0; x < grid.width; x++) {
      if (grid.map[y][x] === PAPER_ROLL) {
        const point = new Point(x, y);
        const connectedRolls = countPaperRollContacts(grid, point);
        if (connectedRolls < 4) {
          part1Count++;
          stack.push(point);
        }
      }
    }
  }

  while (stack.length > 0) {
    const point = stack.pop();

    if (grid.map[point.y][point.x] !== PAPER_ROLL) {
      continue;
    }

    grid.setObject(point, EMPTY);
    part2Count++;

    for (const dir of DIRECTIONS) {
      const neighbor = movePoint(point, dir);
      if (grid.positionContainsObject(neighbor, PAPER_ROLL)) {
        const connectedRolls = countPaperRollContacts(grid, neighbor);
        if (connectedRolls < 4) {
          stack.push(neighbor);
        }
      }
    }
  }

  return [part1Count, part2Count];
}

function calculatePaperRollsOptimized(grid) {
  let part1Count = 0;
  let part2Count = 0;

  const mappedPaperRolls = new Map();
  for (let y = 0; y < grid.height; y++) {
    for (let x = 0; x < grid.width; x++) {
      if (grid.map[y][x] === PAPER_ROLL) {
        const key = `${x},${y}`;
        mappedPaperRolls.set(key, new Point(x, y));
        const connectedRolls = countPaperRollContacts(grid, new Point(x, y));
        if (connectedRolls < 4) {
          part1Count++;
        }
      }
    }
  }

  let lastPassCount = -1;
  while (lastPassCount !== 0) {
    lastPassCount = 0;

    for (const [key, point] of Array.from(mappedPaperRolls)) {
      const connectedRolls = countPaperRollContacts(grid, point);
      if (connectedRolls < 4) {
        grid.setObject(point, EMPTY);
        mappedPaperRolls.delete(key);
        lastPassCount++;
        part2Count++;
      }
    }
  }

  return [part1Count, part2Count];
}

function calculatePaperRolls(grid) {
  let part1Count = 0;
  let part2Count = 0;

  let firstPass = true;
  let lastPassCount = -1;

  while (lastPassCount !== 0) {
    lastPassCount = 0;

    for (let y = 0; y < grid.height; y++) {
      for (let x = 0; x < grid.width; x++) {
        if (grid.map[y][x] !== PAPER_ROLL) {
          continue;
        }
        const connectedRolls = countPaperRollContacts(grid, new Point(x, y));
        if (connectedRolls < 4) {
          if (firstPass) {
            part1Count++;
          } else {
            grid.setObject(new Point(x, y), EMPTY);
            lastPassCount++;
            part2Count++;
          }
        }
      }
    }

    if (firstPass) {
      firstPass = false;
      lastPassCount = -1;
    }
  }

  return [part1Count, part2Count];
}

function day4() {
  const grid = new Grid('day4.part1');

  let start = performance.now();
  const [part1CountQueue, part2CountQueue] = calculatePaperRollsQueue(grid);
  const elapsed = ((performance.now() - start) / 1000).toFixed(4);

  console.log(`\nDay 4:`);
  console.log(`  Part 1: ${part1CountQueue}, Part 2: ${part2CountQueue} (${elapsed}s)`);
}

day4();
